lib/common/memory.c

```diff
@@ -413,3 +413,26 @@ void h2o_append_to_null_terminated_list(void ***list, void *element)
     (*list)[cnt++] = element;
     (*list)[cnt] = NULL;
 }
+
+char *h2o_strerror_r(int err, char *buf, size_t len)
+{
+#ifndef _GNU_SOURCE
+    strerror_r(err, buf, len);
+    return buf;
+#else
+    /**
+     * The GNU-specific strerror_r() returns a pointer to a string containing the error message.
+     * This may be either a pointer to a string that the function stores in  buf,
+     * or a pointer to some (immutable) static string (in which case buf is unused)
+     */
+    return strerror_r(err, buf, len);
+#endif
+}
+
+void h2o_perror(const char *msg)
+{
+    char buf[128];
+
+    h2o_strerror_r(errno, buf, sizeof(buf));
```

Comment:
Thank you for the changes. I believe you need to use the result of the function call here?

Reply:
Yes, sorry for my miss.
Fixed in 82b0800

Reply:
IMO, Recently most systems implements `strerror` using thread local data.
so indeed, `strerror` and `strerror_r` should be both thread safe on these systems. but man page do not assure this.

all these `strerror_r` related seems historical reasons.